src/Runtime/Values.hs

```diff
@@ -15,13 +15,56 @@ import Data.Array
 import Data.List
 import Data.Aeson hiding (Array)
 import GHC.Generics
+import qualified Data.Map.Strict as Map
 
 -- | Representation of a Board in BoGL,
 -- composed of an NxM array of 'Val'
 type Board = Array (Int, Int) Val
 
--- | Evaluation environment
-type EvalEnv = [(Name, Val)]
+
+-- | Typeclass for a runtime environment
+class RuntimeEnv a where
+  -- insert single key/val pair into the env
+  insertEvalEnv :: (String,Val) -> a -> a
+  -- extend an env with a tuple of keys and vals to be combined
```

Comment:
It would be better to use `[(String, Val)]`. This ensures at the type level that we can't introduce a bug where there is a length mismatch on either the `Strings` or the `Vals`.